### PR TITLE
Task/user/push notification

### PR DESCRIPTION
**Pull request by ditansu**
## Review description
Added tests and code to Push-notification subscription/cancel.
It is possible to subscribe to push notification by POST through `user/notification` endpoint with push_token, and `cancel` this one by DELETE `user/notification/:user_id`

**IMPORTANT** 
This PR based on PR #33 therefore, please merge PR #33 **before** begin review this PR. 

## Subscription to push-notification 
To subscribe to notification need send  `POST-request`  to  `user/notification` endpoint from an authorized user. Request have to contain a push_token in request body.

#### Request from `App` to `Server`
POST `http://dev.cocoaheads.ru:8080/user/notification`
**Header** 
```Swift
[
"Authorization" :  "Bearer Xuq5dHHyv0kr1g+jiSitQQ==",
"app-token": "sdf2r34fsdf342vder0(221"
]
```
**Body**
```JSON
{
	"push_token": "sdfsdfsdf223dwd2dsdf23f23f2sfd"
}
```
#### Response from `Server` to `App`
**Body**
```JSON
{
        "id": "1",
        "user_id": "3",
	"push_token": "sdfsdfsdf223dwd2dsdf23f23f2sfd"
}
```

## Cancel 
To cancel push-notification  need send  `DELETE-request`  to  `user/notification/:user_id` endpoint from an authorized user

#### Request from `App` to `Server` 
DELETE `http://dev.cocoaheads.ru:8080/user/notification/3`
**Header** 
```Swift
[
"Authorization" :  "Bearer Xuq5dHHyv0kr1g+jiSitQQ==",
"app-token": "sdf2r34fsdf342vder0(221"
]
```
#### Response from `Server` to `App`
**Body**
```JSON
{
        "status": "OK",
	"message": "Push-notifications registration is canceled"
}
```

### CODE REVIEW

File: Sources/App/Controllers/User/UserController.swift [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
   
-  let droplet: Droplet!
+  let drop: Droplet!
+  let photoConroller: PhotoConroller
   
-  init(droplet: Droplet) {
-    self.droplet = droplet
+  init(drop: Droplet) {
+    self.drop = drop
+    self.photoConroller = PhotoConroller(drop: self.drop)
```
Comment (dannydaddy3) [inline Sources/App/Controllers/User/UserController.swift]: typo PhotoConroller

File: Sources/App/Models/Creator/Creator.swift [hunk 2]
```diff
@@ -46,13 +51,38 @@ final class Creator: Model {
   // sourcery:end
 }
 
+extension Creator: JSONRepresentable {
+
+  func makeJSON() throws -> JSON {
```
Comment (dannydaddy3) [inline Sources/App/Models/Creator/Creator.swift]: Why not generate?

**Review by dannydaddy3 — CHANGES_REQUESTED** [review]

File: Sources/App/Controllers/User/PushNotificationController.swift [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  func subscribe(_ request: Request) throws -> ResponseRepresentable {
+
+    guard
+      let json = request.json,
+      let pushToken = json[Client.Keys.pushToken]?.string
+    else {
+      throw Abort(.badRequest, reason: "Request's body no have pushToken to the push-notification registration")
+    }
+
+    if let client = try Client.returnIfExcists(request: request) {
```
Comment (dannydaddy3) [inline Sources/App/Controllers/User/PushNotificationController.swift]: typo in `Client.returnIfExcists`.  need - `Client.returnIfExists`

File: Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  func testThatPushTokenIsRegistered() throws {
+
+    let user = User()
+    try! user.save()
+
+    let pushToken = String.randomValue
+
+    let body = try! JSON(node: [ "push_token": pushToken ])
```
Comment (dannydaddy3) [inline Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift]: removed space after leading and  before trailing square bracket

Comment (dannydaddy3) [reply Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift]: ))

File: Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    try! subscribeNotification(for: user, with: body).assertStatus(is: .ok)
+
+     let count = try! App.Client.makeQuery()
+      .filter(App.Client.Keys.pushToken, pushToken)
+      .filter(App.Client.Keys.userId, 1).count()
+
+    XCTAssert(count == 1, "Expected count 1, recived count: '\(count)'")
+  }
+
+  func testThatPushTokenIsRegisteredOnlyOneTimeForCertainUser() throws {
```
Comment (dannydaddy3) [inline Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift]: `onlyOnce`

**Review by dannydaddy3 — CHANGES_REQUESTED** [review]

File: Sources/App/Models/Creator/Creator.swift (hunk 2 above)
Comment (ditansu) [reply Sources/App/Models/Creator/Creator.swift]: How to say sourcery to do this: 
```swift
try json.set(Keys.photoUrl, photoURL())
```
??

File: Sources/App/Controllers/User/PushNotificationController.swift (hunk 3 above)
Comment (ditansu) [reply Sources/App/Controllers/User/PushNotificationController.swift]: fixed, thank you!

File: Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift (hunk 4 above)
Comment (ditansu) [reply Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift]: ))) fixed

File: Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift (hunk 5 above)
Comment (ditansu) [reply Tests/AppTests/Controllers/User/PushNotificationControllerTest.swift]: )) fixed!

File: Sources/App/Models/Creator/Creator.swift (hunk 2 above)
Comment (dannydaddy3) [reply Sources/App/Models/Creator/Creator.swift]: Easily :) we're supporting it 
add  `// sourcery: nestedJSONField` above `func photoURL()`
and to ignore appearing this property field in JSON add `// sourcery: ignoreInJSON` above `  var photoUrl: String`

It's described in Wiki for sourcery)

Comment (dannydaddy3) [reply Sources/App/Models/Creator/Creator.swift]: Check it

Comment (ditansu) [reply Sources/App/Models/Creator/Creator.swift]: hm, I'm check it, problem is here:
```Swift
...
struct Keys {
    static let id = "id"
    static let userId = "user_id"
    static let position = "position"
    static let photoUrl = "photo_url"
    static let info = "info"
    static let url = "url"
    static let active = "active"
    static let photoURL = "photo_url"
  }
```
double  `static let photoURL = "photo_url"` there was no have time to resolve it, what is why I didn't use toJSON generation

**Review by dannydaddy3 — APPROVED** [review]

File: Sources/App/Models/Creator/Creator.swift (hunk 2 above)
Comment (ditansu) [reply Sources/App/Models/Creator/Creator.swift]: fixed, now we have photoURL() as warp on nested user.photoURL.

**Pull request merged**